server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java

```diff
@@ -0,0 +1,61 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.adapter.mailbox;
+
+import javax.inject.Inject;
+
+import org.apache.james.core.Username;
+import org.apache.james.mailbox.Authorizator;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.user.api.DelegationStore;
+import org.apache.james.user.api.UsersRepository;
+import org.apache.james.user.api.UsersRepositoryException;
+
+import reactor.core.publisher.Flux;
+
+/**
+ * Authorizator which use an DelegationStore to check if the delegation is allowed
+ */
+public class DelegationStoreAuthorizator implements Authorizator {
+    private final DelegationStore delegationStore;
+    private final UsersRepository usersRepository;
+
+    @Inject
+    public DelegationStoreAuthorizator(DelegationStore delegationStore, UsersRepository usersRepository) {
+        this.delegationStore = delegationStore;
+        this.usersRepository = usersRepository;
+    }
+
+    @Override
+    public AuthorizationState canLoginAsOtherUser(Username userId, Username otherUserId) throws MailboxException {
+        boolean isAuthorized = Flux.from(delegationStore.authorizedUsers(otherUserId)).collectList().block().contains(userId);
```

Comment:
This represent an extra DB read for each and every login attempt. Maybe we need to look up the delegation store if, and only if the two usernames are different?

If they are the same we can skip the delegation and assume ALLOWED here?

Reply:
We already checked that using `AuthenticationAttempt::isDelegation` before deciding to do delegation or not. So I think it is no need for further checking.

Reply:
Ok fair. We will need to keep that in mind for the OIDC delegation stuff, not to call loginAsOtherUser if both users are the same.